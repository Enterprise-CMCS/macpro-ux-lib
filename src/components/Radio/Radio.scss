@use "../../sass/macbis-colors" as c;

.usa-radio__input--tile+[class*=__label] {
  border: 1px solid c.$gray-lighter;
  border-radius: 0.2rem;
}

.usa-radio__input:checked+[class*=__label]::before {
  background-color: c.$primary;
  box-shadow: 0 0 0 2px c.$primary, inset 0 0 0 2px #fff;
}

.usa-radio__input--tile:checked+[class*=__label] {
  background-color: c.$primary-alt-lightest;
  border-color: c.$primary;
}

.usa-radio__input--tile:disabled+[class*=__label] {
  background-color: c.$gray-lightest;
  border-color: c.$gray-lighter;
}

.usa-radio__input:disabled+[class*=__label]::before, .usa-radio__input[aria-disabled=true]+[class*=__label]::before {
  background-color: c.$gray-lighter;
  box-shadow: 0 0 0 2px rgb(27 27 27 / 30%);
}
