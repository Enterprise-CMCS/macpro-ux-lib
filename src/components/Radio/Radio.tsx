import React from "react";

type IntrinsicElements = JSX.IntrinsicElements["input"];

export interface RadioProps extends IntrinsicElements {
  id: string;
  label: string;
  children?: JSX.Element[];
  disabled?: boolean;
  isTile?: boolean;
  tileDescription?: string;
  value?: string;
}

/**
 * Radio Component
 * @param {string}  id                A unique identifier for the input.
 * @param {string}  label             Label text that appears to the right of the Radio.
 * @param {boolean} [checked]         Set the initial checked state.
 * @param {Array}   [children]        An array of child elements to appear when selected.
 * @param {boolean} [disabled]        Sets the Radio to its disabled state.
 * @param {boolean} [isTile]          Use the tile variation of the Radio.
 * @param {string}  [tileDescription] Text that can be used to describe the label in more detail. Activates the tile variation automatically.
 * @param {string}  [value]           Value of the input element.
 */
export const Radio: React.FC<RadioProps> = ({
  children,
  disabled = false,
  id,
  isTile = false,
  label,
  tileDescription,
  value,
  ...rest
}) => {
  return (
    <div className="usa-radio">
      <input
        className={
          isTile
            ? "usa-radio__input usa-radio__input--tile"
            : "usa-radio__input"
        }
        disabled={disabled}
        id={id}
        type="radio"
        value={value}
        {...rest}
      />
      <label className="usa-radio__label" htmlFor={id}>
        {label}
        {isTile && tileDescription && (
          <span className="usa-radio__label-description">
            {tileDescription}
          </span>
        )}
      </label>
    </div>
  );
};
