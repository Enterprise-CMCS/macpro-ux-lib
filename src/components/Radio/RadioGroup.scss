@use "../../sass/macbis-colors" as c;

.border-primary {
  border-color: c.$primary;
}
