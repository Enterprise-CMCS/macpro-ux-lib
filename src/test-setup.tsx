import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
