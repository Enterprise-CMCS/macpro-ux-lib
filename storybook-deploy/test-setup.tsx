import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
