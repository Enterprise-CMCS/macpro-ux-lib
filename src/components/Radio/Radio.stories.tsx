import { ComponentMeta, ComponentStory } from "@storybook/react";
import React from "react";
import { Radio } from "./Radio";

export default {
  title: "Components/Radio",
  component: Radio,
  argTypes: {
    id: {
      description:
        "A unique identifier for the input which must be unique in the whole document. Its purpose is to identify the element when linking. The value is used as the value of the `<label>`'s `for` attribute to link the label with the form control.",
    },
    isTile: {
      description: "Use the tile variation of the Radio.",
    },
    label: {
      description: "Label text that appears to the right of the Radio.",
    },
    name: {
      description:
        "Name of the input element. Radio components are grouped together by `name`.",
    },
    tileDescription: {
      description:
        "Text that can be used to describe the label in more detail. Appears only when `isTile` is set to `true`.",
    },
    value: {
      description:
        "Value of the input element. If you omit the value attribute, the submitted form data assigns the value `on` to the group. For example, if the user clicked on a \"Phone\" option in a Radio grop named `contact` with no `value` set, the resulting form data after submission would be `contact=on`, which isn't helpful. So don't forget to set your value attributes!",
      if: { arg: "isTile" },
    },
  },
  args: {
    disabled: false,
    isTile: false,
    label: "Radio Item",
    tileDescription: "",
  },
  parameters: {
    docs: {
      description: {
        component:
          "A single Radio component, with default and tile variations. Two or more Radio components may be contained in a RadioGroup component in order to manage selected state and show/hide any provided children. See the RadioGroup component for more information.",
      },
    },
  },
} as ComponentMeta<typeof Radio>;

const Template: ComponentStory<typeof Radio> = ({ ...rest }) => (
  <Radio {...rest} />
);

export const Default = Template.bind({});
Default.args = {
  id: "radio-default-1",
  name: "radio-default",
  value: "radio-default-1",
};

export const Disabled = Template.bind({});
Disabled.args = {
  id: "radio-disabled-1",
  name: "radio-disabled",
  value: "radio-disabled-1",
  disabled: true,
};

export const Tile = Template.bind({});
Tile.args = {
  id: "radio-tile-1",
  isTile: true,
  name: "radio-tile",
  value: "radio-tile-1",
};

export const TileWithDescription = Template.bind({});
TileWithDescription.args = {
  id: "radio-desc-tile-1",
  isTile: true,
  name: "radio-tile-desc",
  tileDescription: "This is a tile description.",
  value: "radio-desc-tile-1",
};

export const DisabledTile = Template.bind({});
DisabledTile.args = {
  id: "radio-desc-tile-disabled",
  isTile: true,
  disabled: true,
  tileDescription: "This is a tile description.",
  value: "radio-desc-tile-disabled",
};
