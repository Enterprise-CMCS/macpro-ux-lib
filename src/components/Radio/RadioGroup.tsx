import React, { ReactElement } from "react";
import { Radio, RadioProps } from "../Radio/Radio";

export interface RadioGroupProps {
  groupName: string;
  radioProps: RadioProps[];
}

export const RadioGroup: React.FC<RadioGroupProps> = ({ groupName, radioProps }) => {
  // Track which Radio is currently selected in order to show/hide child elements:
  const [selected, setSelected] = React.useState("");
  const handleChange = (e: React.ChangeEvent) => {
    setSelected(e.target.id);
  };

  // Create Radio components for the RadioGroup based `radioProps`:
  const radioButtons = (radioProps: RadioProps[]): ReactElement[] => {
    let radios: ReactElement[] = [];

    for (var props of radioProps) {
      radios.push(
        <div key={`${groupName}-${props.id}`}>
          <Radio
            key={props.id}
            onChange={handleChange}
            name={groupName}
            {...props}
          />
          {props.id === selected && props.children && (
            <div className="border-left-05 border-primary margin-left-1 padding-left-205">
              {props.children}
            </div>
          )}
        </div>
      );
    }

    return radios;
  };

  return <>{radioButtons(radioProps)}</>;
};
