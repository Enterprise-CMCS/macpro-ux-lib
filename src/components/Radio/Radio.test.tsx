import { Radio } from "../Radio/Radio";
import { render, screen } from "../../test-setup";
import fireEvent from "@testing-library/user-event";
import React from "react";

describe("Radio", () => {
  describe("default", () => {
    beforeEach(() => {
      render(
        <form data-testid="test-form">
          <Radio
            id="radio-default"
            name="default-group"
            label="Default Radio"
            value="choice-1"
          />
        </form>
      );
    });

    it("is unselected by default", () => {
      const r = screen.getByRole("radio");
      expect(r).not.toBeChecked();
    });

    it("becomes selected on click", () => {
      const r = screen.getByRole("radio");
      fireEvent.click(r);
      expect(r).toBeChecked();
    });

    it("has the expected label", () => {
      const r = screen.getByLabelText("Default Radio");
      expect(r).toBeInTheDocument();
    });

    it("is not a tile by default", () => {
      const r = screen.getByRole("radio");
      expect(r).not.toHaveClass("usa-checkbox__input--tile");
    });

    it("has the expected ID", () => {
      const r = screen.getByRole("radio");
      expect(r.getAttribute("id")).toBe("radio-default");
    });

    it("has the expected name attribute", () => {
      const r = screen.getByRole("radio");
      expect(r.getAttribute("name")).toBe("default-group");
    });
  });
});

describe("Tile", () => {
  beforeEach(() => {
    render(
      <form data-testid="test-form">
        <Radio
          id="radio-tile"
          isTile={true}
          label="Tile Radio"
          name="tile-group"
          tileDescription="This is the tile version of the Radio component."
        />
      </form>
    );
  });

  it("renders as a tile", () => {
    const r = screen.getByRole("radio");
    expect(r).toHaveClass("usa-radio__input--tile");
  });

  it("is unchecked by default", () => {
    const r = screen.getByRole("radio");
    expect(r).not.toBeChecked();
  });

  it("becomes checked on click", () => {
    const r = screen.getByRole("radio");
    fireEvent.click(r);
    expect(r).toBeChecked();
  });

  it("renders and has the expected label", () => {
    const r = screen.getByRole("radio");
    expect(r).toBeInTheDocument();
  });

  it("has the expected ID", () => {
    const r = screen.getByRole("radio");
    expect(r.getAttribute("id")).toBe("radio-tile");
  });

  it("has the expected name attribute", () => {
    const r = screen.getByRole("radio");
    expect(r.getAttribute("name")).toBe("tile-group");
  });

  it("has the expected description text", () => {
    const testForm = screen.getByTestId("test-form");
    expect(testForm).toHaveTextContent(
      "This is the tile version of the Radio component."
    );
  });
});
