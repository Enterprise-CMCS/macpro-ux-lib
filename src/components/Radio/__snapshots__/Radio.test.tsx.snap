// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Snapshot tests renders default Radio as expected 1`] = `
<div>
  <div
    class="usa-radio"
  >
    <input
      class="usa-radio__input"
      id="default-radio"
      name="default-group"
      type="radio"
      value="default-radio"
    />
    <label
      class="usa-radio__label"
      for="default-radio"
    >
      Default Radio
    </label>
  </div>
</div>
`;

exports[`Radio Snapshot tests renders disabled Radio as expected 1`] = `
<div>
  <div
    class="usa-radio"
  >
    <input
      class="usa-radio__input"
      disabled=""
      id="default-radio"
      name="default-group"
      type="radio"
      value="default-radio"
    />
    <label
      class="usa-radio__label"
      for="default-radio"
    >
      Default Radio
    </label>
  </div>
</div>
`;

exports[`Radio Snapshot tests renders disabled tile Radio as expected 1`] = `
<div>
  <div
    class="usa-radio"
  >
    <input
      class="usa-radio__input usa-radio__input--tile"
      disabled=""
      id="tile-radio"
      name="tile-group"
      type="radio"
      value="tile-radio"
    />
    <label
      class="usa-radio__label"
      for="tile-radio"
    >
      Tile Radio
      <span
        class="usa-radio__label-description"
      >
        This is the tile Radio description.
      </span>
    </label>
  </div>
</div>
`;

exports[`Radio Snapshot tests renders tile Radio as expected 1`] = `
<div>
  <div
    class="usa-radio"
  >
    <input
      class="usa-radio__input usa-radio__input--tile"
      id="tile-radio"
      name="tile-group"
      type="radio"
      value="tile-radio"
    />
    <label
      class="usa-radio__label"
      for="tile-radio"
    >
      Tile Radio
      <span
        class="usa-radio__label-description"
      >
        This is the tile Radio description.
      </span>
    </label>
  </div>
</div>
`;
